syntax = "proto3";

message TimeVal {
	uint32 tv_sec = 1;// store time elapse since Jan 1, 1970 00:00 in seconds
	uint32 tv_usec = 2; // store time elapse since Jan 1, 1970 00:00 in microseconds
}

message SettimeofdayRequest {
	TimeVal timeval = 1;
}

message SettimeofdayResponse {
	sint32 return_value = 1; //stores if function call was succesful or not
	sint32 errno_alt_alt = 2; //stores error no., if fault occurs
}

message GettimeofdayRequest {
    int32 stub = 1;
}

message GettimeofdayResponse{
	//planning to scrap timezone completely, as it is obsolete
	sint32 return_value = 1;//stores if function call was succesful or not
	sint32 errno_alt_alt = 2;
	TimeVal timeval = 3;
}

message Request {
	oneof RequestFunc {
		SettimeofdayRequest settimeofday_request = 1;
		GettimeofdayRequest gettimeofday_request = 2;
	}
}

message Response {
	oneof ResponseFunc {
		SettimeofdayResponse settimeofday_response = 1;
		GettimeofdayResponse gettimeofday_response = 2;
	}
}


message xRPCMessage {
	oneof MessageType {
		Request request   = 1;
		Response response = 2;
	}
	TimeVal time_stamp    = 3;

	// oneof is simlar to the union structure in C
	// the *_case variable stores the selection of which type
	/*
	oneof RequestType { 
		SettimeofdayRequest settimeofday_request = 10;
		GettimeofdayRequest gettimeofday_request = 11;

		SettimeofdayResponse settimeofday_response = 1003;
		GettimeofdayResponse gettimeofday_response = 1004;
	}
	*/
}
